app/services/email_verification_code_service.rb

```diff
@@ -0,0 +1,69 @@
+class EmailVerificationCodeService
+  VERIFICATION_TYPES = [:ctc_intake, :gyr_login]
+  def initialize(email_address: , locale: :en, visitor_id: , client_id: nil, verification_type: :gyr_login)
+    raise(ArgumentError, "Unsupported verification type: #{verification_type}") unless VERIFICATION_TYPES.include? verification_type
+
+    @verification_type = verification_type.to_sym
+    @email_address = email_address
+    @locale = locale
+    @visitor_id = visitor_id
+    @client_id = client_id
+  end
+
+  def request_code
+    can_send_code? ? send_verification_code : send_no_match_message
+  end
+
+  private
+
+  def can_send_code?
+    case @verification_type
+    when :ctc_intake
+      true
+    when :gyr_login
+      ClientLoginsService.accessible_intakes.where(email_address: @email_address).or(
+      ClientLoginsService.accessible_intakes.where(spouse_email_address: @email_address)).exists?
+    end
+  end
+
+  # sends an email synchronously using deliver_now so that we can store the mailgun ID,
+  # so best to send this from a background job.
+  def send_verification_code
+    verification_code, access_token = generate_verification_code
+    mailer_response = VerificationCodeMailer.with(
+      to: @email_address,
+      verification_code: verification_code,
+      locale: @locale,
+      verification_type: @verification_type
+    ).with_code.deliver_now
```

Comment:
The biggest change between the previous implementation and this implementation is that I'm using deliver_now for the email so that I can get the message ID that we use for detecting mailgun webhooks and update the status of the emails. Also, Since we always call this service from a background job, using deliver_later for sending the messages was queuing this job, and then queuing another job to send the email. Seemed like we could make that more efficient by sending the email in the job itself.